[IMDCE] Remove dead hierpath and annotations in IMDCE

uenoku opened the pull request:
This PR adds functionality to find and remove dead hierpath in IMDCE. There are discardable annotations (e.g. OMTracker) which use hierpath and prevent IMDCE from deleting instances with inner symbols. This PR imports some of InnerSymbolDCE logic into IMDCE.

mikeurbach commented on lib/Dialect/FIRRTL/Transforms/IMDeadCodeElim.cpp:
```diff
@@ -524,8 +634,8 @@ void IMDeadCodeElimPass::rewriteModuleSignature(FModuleOp module) {
       if (argument.getType().isa<RefType>())
         continue;
 
-      // Ok, this port is used only within its defined module. So we can replace
```
Seems like the auto-formatter disagreed with some comment line breaking. Was it wrong before?

uenoku replied:
Hmm, I think I accidentally relayout in a different way, will fix. Thanks!

mikeurbach commented on lib/Dialect/FIRRTL/Transforms/IMDeadCodeElim.cpp:
```diff
@@ -287,8 +381,28 @@ void IMDeadCodeElimPass::runOnOperation() {
+      if (!hierPathSym)
+        return false;
+      auto op =
+          symbolTable->template lookup<hw::HierPathOp>(hierPathSym.getAttr());
```
I seem to recall something about this `template` usage tripping up MSVC, but I'm not sure.

uenoku replied:
Thanks, testing here https://github.com/llvm/circt/actions/runs/5070937958

mikeurbach commented on lib/Dialect/FIRRTL/Transforms/IMDeadCodeElim.cpp:
```diff
@@ -116,9 +122,41 @@ struct IMDeadCodeElimPass : public IMDeadCodeElimBase<IMDeadCodeElimPass> {
+    for (auto hierPathOp : instanceToHierpath[instance.getInnerRef()])
+      markAlive(hierPathOp);
+  }
+  // If the output port is alive, mark the instnace as alive. Propagate the
```
nit: I know this comment already existed, but it seems like the `instnace` typo was carried over. If you're making any other fixes to this PR, maybe update this while you're at it?

uenoku replied:
Yes, this comment is not accurate. I moved the function from the different place 😔

mikeurbach reviewed (commented):
Nice, I am not super familiar with this pass, but I think it looks good to me. Just left some minor feedback.

prithayan commented on lib/Dialect/FIRRTL/Transforms/IMDeadCodeElim.cpp:
```diff
@@ -259,8 +298,63 @@ void IMDeadCodeElimPass::forwardConstantOutputPort(FModuleOp module) {
+      // caught the error. Conservatively mark the symbol as alive.
+      if (hierPath.isPublic() || namePath.size() <= 1 ||
+          namePath.back().isa<hw::InnerRefAttr>()) {
+        markAlive(hierPath);
```
Why are hierpaths to modules not marked alive ?

uenoku replied:
Hirepaths to modules is exactly what I want to remove in this PR, so liveness of such hierpaths are propagated in the main loop.

prithayan commented on lib/Dialect/FIRRTL/Transforms/IMDeadCodeElim.cpp:
```diff
@@ -259,8 +298,63 @@ void IMDeadCodeElimPass::forwardConstantOutputPort(FModuleOp module) {
+    for (NamedAttribute namedAttr : op->getAttrs()) {
+      namedAttr.getValue().walk([&](Attribute subAttr) {
+        if (auto innerRef = dyn_cast<hw::InnerRefAttr>(subAttr)) {
+          if (auto instance = dyn_cast_or_null<firrtl::InstanceOp>(
```
why is only the instance being marked alive, what if its an `InnerRef` to a  wire, should we mark the wire also live ?

uenoku replied:
Actually declarations(wire/node/reg) with inner symbols are marked as alive in `markDeclaration` function. I want to keep this PR to delete only dead hierpath(and associated instances) so this PR doesn't handle dead inner symbols other than instances. Ideally IMDCE should inner symbols on declarations too but I think it would require extra work.

mikeurbach replied:
Makes sense, maybe we can file an issue for a future enhancement? I can imagine the same situation with a declaration other than instance.

uenoku commented on lib/Dialect/FIRRTL/Transforms/IMDeadCodeElim.cpp:
```diff
@@ -259,8 +298,63 @@ void IMDeadCodeElimPass::forwardConstantOutputPort(FModuleOp module) {
+      auto namePath = hierPath.getNamepath().getValue();
+      // If the hierpath is public or ill-formed, the verifier should have
+      // caught the error. Conservatively mark the symbol as alive.
+      if (hierPath.isPublic() || namePath.size() <= 1 ||
```
I want to pattern match the heierpath with `[..., @foo::@bar, @moduleName]`, so we want to reject other patterns (not sure three still exists such a hierpath though).

uenoku commented:
> High level comments, what's the performance implication of computing the innerRefNamespace

I tesed large cores, but it increases 1% of entire execution time, so I'd consider acceptable.

uenoku commented:
I updated the PR not to introduce the notion of "discardable/weak/value tap annotation" which requires more discussion (https://github.com/llvm/circt/pull/5240). Instead, I modified OMIRTracker annotations to contain tracker types and IMDCE pass will determine whether it's valid to drop the tracker annotations. Eventually the logic should be replaced with "circt.discardbale" whatever.

dtzSiFive commented:
Computing IRN should be fairly fast, esp if there are many cores available, but good question.  We presently compute it as part of verification between passes (that don't preserve everything), although there's no re-use from the verifier-computed version.

dtzSiFive commented on lib/Dialect/FIRRTL/Transforms/IMDeadCodeElim.cpp:
```diff
@@ -259,8 +318,60 @@ void IMDeadCodeElimPass::forwardConstantOutputPort(FModuleOp module) {
+
+  // Walk attributes and find unknown uses of inner symbols or hierpaths.
+  getOperation().walk([&](Operation *op) {
+    if (isa<FModuleOp>(op)) // Port or module annoations are ok.
```
```suggestion
    if (isa<FModuleOp>(op)) // Port or module annotations are ok.
```

uenoku replied:
Thanks!

darthscsi commented on include/circt/Dialect/FIRRTL/Passes.td:
```diff
@@ -117,7 +117,7 @@ def RemoveUnusedPorts : Pass<"firrtl-remove-unused-ports", "firrtl::CircuitOp">
   ];
 }
 
-def IMDeadCodeElim : Pass<"firrtl-imdeadcodeelim", "firrtl::CircuitOp"> {
```
I think hierpath are now in the circuit namespace.

uenoku replied:
I followed the implementation of InnerSymbolDCE that checks all uses of inner symbols not only within circuit op.

https://github.com/llvm/circt/blob/a6c8cb3b14ae1b3f31ce28262c2be54b6decac57/lib/Dialect/FIRRTL/Transforms/InnerSymbolDCE.cpp#L107-L109 

If it's assumed that there is no user of inner symbol outer circuit op, it makes sense to me to make these passes circuit op.

darthscsi commented on lib/Dialect/FIRRTL/Transforms/IMDeadCodeElim.cpp:
```diff
@@ -42,6 +44,16 @@ static bool isDeletableDeclaration(Operation *op) {
+  if (!anno.isClass(omirTrackerAnnoClass))
+    return false;
+
+  auto tpe = anno.getMember<StringAttr>("type");
```
This is NOT acceptable.  No pass should be interpreting annotations except passes explicitly dealing with those annotations.
